### ISSUE TITLE
Multi Wrapper setup doesn't take into account the policy + ( error message wrong )

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1092]
In GitLab by @eferre1 on Jul 21, 2023, 15:20

Hello @dbeltrankyl and @kinow,

### Autosubmit Version

v3.15.0b

### Expid affected

`a5ls`

### Which task has issues? Where is the log?

The issue is with the wrappers, and also the submission of the individual jobs (this part was already reported on #1062).

### Summary

I'm currently running a multi-member multi-startdate (2 of each) experiment in MeluXina to test the working of the horizontal wrappers. I set up a multi-wrapper with a vertical one for the SIM jobs to mimic the expected use in general.

The issue arises when the first SIM job finishes (there are 4 running in parallel in their respective vertical wrappers due to the fast queuing of this machine, but this won't be the usual behavior), and then there are 2 new jobs (both in the horizontal wrapper) available. The wrappers have this settings:

```
[wrapper]
TYPE = multi
WRAPPER_LIST = SIM_WRAPPER,CMOR_WRAPPER

[SIM_WRAPPER]
TYPE = vertical
JOBS_IN_WRAPPER = SIM
MIN_WRAPPED = 3
MAX_WRAPPED = 10
POLICY = flexible #default is flexible. Values: flexible,strict,mixed

[CMOR_WRAPPER]
TYPE = horizontal
JOBS_IN_WRAPPER = CMORATM&CMOROCE&SAVEIC&POST
MIN_WRAPPED = 4
MAX_WRAPPED = 20
POLICY = mixed #default is flexible. Values: flexible,strict,mixed
```

Which should make these 2 jobs wait for another 2 before being submitted together with a wrapper since they have the mixed policy. But for some reason, the log says that the wrapper has flexible policy and submits them independently. This crashes autosubmit since the SAVEIC job has the following header:

```
###############################################################################
#                   SAVEIC a5ls EXPERIMENT
###############################################################################
#
#SBATCH --qos=default
#SBATCH -A p200125
#
#SBATCH --cpus-per-task=12
#SBATCH --tasks-per-node=128
#
#SBATCH -n 1
#SBATCH -t 00:40:00
#SBATCH -J a5ls_19900201_fc00_1_SAVEIC
#SBATCH --output=/project/home/p200125/u100498/a5ls/LOG_a5ls/a5ls_19900201_fc00_1_SAVEIC.cmd.out.0
#SBATCH --error=/project/home/p200125/u100498/a5ls/LOG_a5ls/a5ls_19900201_fc00_1_SAVEIC.cmd.err.0
#SBATCH -N 1
#SBATCH --hint=nomultithread
#
#
###############################################################################
```

And it conflicts the `--cpus-per-task`, the `--tasks-per.node` and the reserved nodes. This crash comes from the issue mentioned (#1062), where it introduces some wrong lines on the job header that were for the wrappers settings.

But the main objective of this issue, is to discern why it is considering the horizontal wrapper as having `flexible` policy.

### Steps to reproduce

Copying and running the `a5ls` experiment should show the issue (but it requires access to MeluXina). Another way would be to generate a multi-member multi-startdate experiment, with a multi-wrapper configuration (with flexible and mixed policies) and that the mixed policy one gets less jobs than the minimum as ready to be submitted. Then, if it behaves as in this case, it will say that there is a deadlock and that due to the flexible policy it will submit the jobs independently.

### What is the current bug behavior?

Here I add the output of the `autosubmit run` command, the crash as I mention is due to another issue, and the problem at hand is the line regarding the flexible policy.

```
717 of 724 jobs remaining (12:45)
Job a5ls_20000201_fc00_1_SIM is RUNNING
Job a5ls_20000201_fc01_1_SIM is RUNNING
Job a5ls_19900201_fc01_1_SIM is RUNNING
Job a5ls_19900201_fc00_1_SIM is COMPLETED
Job a5ls_19900201_fc00_1_SIM finished at 2023-07-21 12:45:59
Wrapper policy is set to flexible and there is a deadlock, As will submit the jobs sequentally
[ERROR] Trace: Command /project/home/p200125/u100498/a5ls/LOG_a5ls/submit_meluxina.sh in meluxina_autosubmit warning: sbatch: error: Batch job submission failed: Requested node configuration is not available


 [CRITICAL] Check that meluxina platform has set the correct scheduler. Sections that could be affected: SAVEIC&CMOROCE [eCode=7014]
More info at https://autosubmit.readthedocs.io/en/v3.15.0/faq.html
```

### What is the expected correct behavior?

As I mention, the crash is not (fully) related to this issue, only the part affecting the wrappers policies.

The expected behavior would be for the CMOROCE and SAVEIC jobs to wait until a wrapper of the required dimensions can be submitted.

### Relevant logs and/or screenshots

Everything is already added on the other parts of the issue, but if you think any other log can be useful I can add them later.

### Any other relevant information

After talking about this with Dani, we were thinking that the problem comes from the multi-wrapper setting, that for some reason is not taking into account the policy, but it would be needed to discern if it is taking the default (flexible) for both wrappers, or not, since one of them is already using it. So he mentioned that it could be worth to test if setting both wrappers to mixed it detects that policy for the vertical one. Or maybe (my thoughts), that there is only taking into account the first policy set up in the conf file, and then setting it for all the wrappers.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 26, 2023, 09:51

Hello @eferre1 ,

I think the issue is the in-line comment, in Autosubmit it is read

`CMOR_WRAPPER=u'mixed #default is flexible. values: flexible,strict,mixed'`

When I put the comment outside the line it works, sorry for not noticing it earlier I think this happened before but in Yaml it doesn't happen so I forgot 

```
[CMOR_WRAPPER]
TYPE = horizontal
JOBS_IN_WRAPPER = CMORATM&CMOROCE&SAVEIC&POST
MIN_WRAPPED = 4
MAX_WRAPPED = 20
POLICY = mixed #default is flexible. Values: flexible,strict,mixed
```

to

```
[CMOR_WRAPPER]
TYPE = horizontal
JOBS_IN_WRAPPER = CMORATM&CMOROCE&SAVEIC&POST
MIN_WRAPPED = 4
MAX_WRAPPED = 20
#default is flexible. Values: flexible,strict,mixed
POLICY = mixed 
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 26, 2023, 10:26

Ok, I'll test it, thanks, I'm not sure we noticed in this same field, but now that you mention it, I remember to have seen something like this somewhere, about a comment influencing the autosubmit variable.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 26, 2023, 11:37

Hi, now it seems to be working (haven't reproduced the same exact situation, but reached a waiting point), but the message when waiting for enough jobs to be wrapped doesn't look completely right (since the horizontal min is 4, not 1, and the comparison doesn't make sense that way):

```
[ERROR] Wrapper policy is set to MIXED and there are not enough jobs to form a wrapper.[wrappeable:4 < defined_min:1] [wrappeable_h:3 < defined_min_h:1]|[wrappeable_v:1 < defined_min_v:1] waiting until the wrapper can be formed.[eCode=6013]
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 26, 2023, 12:06

Another thing that I noticed, but that might be my fault since I'm not sure if I'm using the correct parameter, is that the wrappers are being formed with 20 jobs each, disregarding the setting on the `jobs.conf`:

```
[CMOROCE]
THREADS = 8
[CMORATM]
THREADS = 8
[SAVEIC]
THREADS = 12
```

Since it is set up to use only one node (128 processors), I was thinking that this would limit the number of jobs per wrapper (since it is from here where it reads the amount of processors to ask for when submitted individually). But in this case, it looks like it is not taking this into account, here you can see the header of a (20 jobs) SAVEIC wrapper:

```
###############################################################################
#              a5ls_ASThread_16903623302732_20_20
###############################################################################
#
#SBATCH -J a5ls_ASThread_16903623302732_20_20
#SBATCH --qos=default
#SBATCH -A p200125
#SBATCH --output=a5ls_ASThread_16903623302732_20_20.out
#SBATCH --error=a5ls_ASThread_16903623302732_20_20.err
#SBATCH -t 00:40:00
#SBATCH --cpus-per-task=12
#SBATCH -n 20
#
#SBATCH -N 1
            #SBATCH --hint=nomultithread
#
```

It is asking for 20 processors, and then the `--cpus-per-task=12` is added from the threads parameter I guess.

It didn't cause a problem, since the nodes are big and the jobs could run, but I wanted to know how this should be set up to limit the number of jobs in a wrapper, apart from setting a lower max limit, since the jobs will not always be the same kind in this wrapper, and the idea was to limit the different jobs depending on its requirements.

So with the current settings, it should have been a max of 10 SAVEIC jobs or 16 CMOR ones, or any combination of both that doesn't use more than 128 processors. Should it be `PROCESSORS` instead of `THREADS` (I think it was like this due to the later usage of the parameters in the jobs)? Any idea @dbeltrankyl?

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 26, 2023, 15:22

mentioned in issue auto-ecearth3#1822

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 27, 2023, 08:33

> Hi, now it seems to be working (haven't reproduced the same exact situation, but reached a waiting point), but the message when waiting for enough jobs to be wrapped doesn't look completely right (since the horizontal min is 4, not 1, and the comparison doesn't make sense that way):

Seems that message is wrong for mixed, yesterday I checked it and is taking a wrong value but only in the final message, so the wrapper itself should work as intended.  I'm fixing the message

> It is asking for 20 processors, and then the `--cpus-per-task=12` is added from the threads parameter I guess.

Yes, it takes the highest of the ones inside the section 

> So with the current settings, it should have been a max of 10 SAVEIC jobs or 16 CMOR ones, or any combination of both that doesn't use more than 128 processors. Should it be `PROCESSORS` instead of `THREADS` (I think it was like this due to the later usage of the parameters in the jobs)? Any idea @dbeltrankyl?


I'm not sure if I understood the question, Do you mean that `CMORATM&CMOROCE&SAVEIC&POST` will not be packed together, instead there could be 3 cmoratm 0 cmoroce 0 saveic 6 post? per example? 

apart of forcing the max_wrapper and min_wrapper values, I don't think you can set a maximum of each kind, this maybe could be a future improvement but I don't think it belongs to 3.15 due time constrains and because this development would be totally different than in 4.0 due the changes on how to use multi wrappers etc 

We could add a setting to specify the max by section alongside the general one and the horizontal and vertical ones.

For now, the only things that determine the capacity of the wrapper are: 

* MAX/MIN in wrapper section in a general basis for all the wrapper 
* Processors(job) and MAX_PROCESSORS(platform)
* wallclock and Total_wallclock(platform)

Maybe you can play with these values, artificially adding a higher wallclock?

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 27, 2023, 08:39

> Do you mean that `CMORATM&CMOROCE&SAVEIC&POST` will not be packed together, instead there could be 3 cmoratm 0 cmoroce 0 saveic 6 post? per example?

Yes, the normal and expected behavior would be to have 1 of each (or maybe 2 or 3, but the same amount of each). But in case of failures or unexpected it could happen that there are more of one kind waiting, and the idea would be to restrict them depending on their needs. From what you mention I guess that setting the processors on each job (instead of threads) should do the trick if we need it.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 27, 2023, 08:51

Originally the wrappers were thought about being always the same size and using the same resources. After some developments now more flexible, so what you mention is now possible but you can't fine tune it. 

So yes, the processors mechanic is the way of doing it. 

However, I will open an issue for As 4.0+ to allow this fine tune without having to modify the jobs itself so it could be used in others workflows or ecearth 4

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 27, 2023, 10:38

Ok, that makes sense, that fine tuning is as always not fully needed, but useful to minimize the resources consumption. As you say, I agree that it is better to keep it for next releases of autosubmit 4.

By the way, I tested the workflow with the changes, and it worked fine once it started to have waiting jobs, didn't send them individually but waited for more to submit the wrappers, so that is exactly what we wanted. You can close this issue. Thanks.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 27, 2023, 18:42

@dbeltrankyl is `MAX_PROCESSORS` only taking into account the sum of all `PROCESSORS` in all wrapped tasks, or the result of `SUM(PROCESSORS*TASKS*CPUS_PER_TASK)`?

I am saying that because, in the case that, let's say, you have a QOS that allows maximum 2,400 cores, the following job will not run: `srun -n 2400 -ntasks-per-node=12`.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jul 27, 2023, 18:48

About the mix of `--ntasks-per-node` and `--cpus-per-task` we were thinking about possible solutions and it's not easy.

For `salloc`, it doesn't matter if you ask 4 processors, with 2 tasks per node (2 nodes) or 2 processors, with 1 task per node (2 nodes). Or if you ask 2 processors, with 48 cpus per task (2 nodes), or 96 processors, with 1 cpu per task (2 nodes). It's a matter of doing a multiplication and we could translate all the TASK specifications into a global processor allocation.

The problem here is that we use the `TASKS` and `THREADS` values in the templates itself and changing these values can lead to unexpected results

In [auto-ec-earth3](https://earth.bsc.es/gitlab/es/auto-ecearth3/-/wikis/PROCESSORS-TASKS-THREADS-definition) we really only use `THREADS` and only in `SAVE_IC` so we are not in danger, but we have to take not of this

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 28, 2023, 09:16

> @dbeltrankyl is `MAX_PROCESSORS` only taking into account the sum of all `PROCESSORS` in all wrapped tasks, or the result of `SUM(PROCESSORS*TASKS*CPUS_PER_TASK)`?

Only processors
